File: src/classes/datas/config/server-config.ts
'use strict'

import { Account } from './account'
import { Repository } from './repository'

export class ServerConfig {

    is_loaded: boolean

    enable_this_device: boolean

    device: string

    is_local_only_access: boolean

    address: string

    enable_tls: boolean

    tls_cert_file: string

    tls_key_file: string

    open_directory_command: string

    open_file_command: string

    urlog_timeout: string

    urlog_useragent: string

    upload_size_limit_month: Number

    user_data_directory: string

    repositories: Array<Repository>

    accounts: Array<Account>

    mi_notification_public_key: string

    mi_notification_private_key: string

    async clone(): Promise<ServerConfig> {
        const server_config = new ServerConfig()
        server_config.is_loaded = this.is_loaded
        server_config.enable_this_device = this.enable_this_device
        server_config.device = this.device
        server_config.is_local_only_access = this.is_local_only_access
        server_config.address = this.address
        server_config.enable_tls = this.enable_tls
        server_config.tls_cert_file = this.tls_cert_file
        server_config.tls_key_file = this.tls_key_file
        server_config.open_directory_command = this.open_directory_command
        server_config.open_file_command = this.open_file_command
        server_config.urlog_timeout = this.urlog_timeout
        server_config.urlog_useragent = this.urlog_useragent
        server_config.upload_size_limit_month = this.upload_size_limit_month
        server_config.user_data_directory = this.user_data_directory
        server_config.repositories = this.repositories
        server_config.accounts = this.accounts
        server_config.mi_notification_public_key = this.mi_notification_public_key
        server_config.mi_notification_private_key = this.mi_notification_private_key
        return server_config
    }

    constructor() {
        this.is_loaded = false
        this.enable_this_device = false
        this.device = ""
        this.is_local_only_access = true
        this.address = "8888"
        this.enable_tls = false
        this.tls_cert_file = ""
        this.tls_key_file = ""
        this.open_directory_command = ""
        this.open_file_command = ""
        this.urlog_timeout = "10s"
        this.urlog_useragent = ""
        this.upload_size_limit_month = 0
        this.user_data_directory = ""
        this.mi_notification_public_key = ""
        this.mi_notification_private_key = ""
        this.repositories = new Array<Repository>()
        this.accounts = new Array<Account>()
    }

}

File: src/server/gkill/dao/server_config/server_config.go
// Package server_config はサーバ設定のDAO。
package server_config

import (
	"time"

	"github.com/mt3hr/gkill/src/server/gkill/dao/account"
	"github.com/mt3hr/gkill/src/server/gkill/dao/user_config"
)

// 待受の既定値。初回起動ブロック（gkill_server_api.go）と DAO の既定マップ
// （server_config_dao_sqlite3_impl.go の serverConfigDefaultValue）が両方ここを引く。
//
// 既定はループバック限定にする。人生記録を保存するアプリなので、何も知らない利用者が
// 起動しただけで LAN の第三者から届く状態にしない。他の端末から使うときは
// サーバ設定画面でアドレスと「ローカルアクセスのみ許可」を明示的に開く。
// 2026-09-14 まで初回起動ブロックは ":9999"（全インターフェース）+ IsLocalOnlyAccess=false を
// リテラルで書いており、DAO 側の既定（IsLocalOnlyAccess=true）と食い違っていた。
// 既定を2箇所に書くとこうなるので、ここ以外に既定値を書かないこと。
const (
	// DefaultListenAddress は待受アドレスの既定。ホスト部を空（":9999"）にすると全インターフェースで待つ。
	DefaultListenAddress = "127.0.0.1:9999"
	// DefaultIsLocalOnlyAccess は「ローカルアクセスのみ許可」の既定。
	// 待受をループバックに限定していても、--address で bind を広げたときの2枚目の防御線になる。
	DefaultIsLocalOnlyAccess = true
)

type ServerConfig struct {
	EnableThisDevice bool `json:"enable_this_device"`

	Device string `json:"device"`

	IsLocalOnlyAccess bool `json:"is_local_only_access"`

	Address string `json:"address"`

	EnableTLS bool `json:"enable_tls"`

	TLSCertFile string `json:"tls_cert_file"`

	TLSKeyFile string `json:"tls_key_file"`

	OpenDirectoryCommand string `json:"open_directory_command"`

	OpenFileCommand string `json:"open_file_command"`

	URLogTimeout time.Duration `json:"urlog_timeout"`

	URLogUserAgent string `json:"urlog_useragent"`

	UploadSizeLimitMonth int `json:"upload_size_limit_month"`

	UserDataDirectory string `json:"user_data_directory"`

	GkillNotificationPublicKey string `json:"gkill_notification_public_key"`

	GkillNotificationPrivateKey string `json:"gkill_notification_private_key"`

	UseGkillNotification bool `json:"use_gkill_notification"`

	GoogleMapAPIKey string `json:"google_map_api_key"`

	LanHostname string `json:"lan_hostname"`

	GlobalHostname string `json:"global_hostname"`

	Repositories []*user_config.Repository `json:"repositories"`

	Accounts []*account.Account `json:"accounts"`
}
